### PR TITLE
Add intellisense support for extension types.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1585]: needed this for reasonable ide recovery.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2242]: the policy is this:

in C# 13, we always treat `implicit extension` as the start of an extension type.  Prior to C# 13, we take the approach of ensuring any prior legal parses stay legal.  But if it would not be legal, we treat it as an extension (with an error that they're not available in your version).

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs:106]: renamed this member to not be such a mouthful.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/BoolKeywordRecommender.cs
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/BoolKeywordRecommender.cs:27]: sorted these for consistency in a lot of our recommender types.  Added `IsExtensionForTypeContext` check in all of them.

File: src/Features/Core/Portable/Shared/Extensions/ISymbolExtensions_2.cs
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Shared/Extensions/ISymbolExtensions_2.cs:88]: we may want a special glyph for extension types.  will be a UX call. this unblocks us from crashing for now.

File: src/EditorFeatures/CSharpTest2/Recommendations/ExplicitKeywordRecommenderTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/ExplicitKeywordRecommenderTests.cs:16]: lots of 'implicit' and 'explicit' keyword tests changed, now that 'implict/explicit' is not just for conversion operators but for implicit/explitic extension types as well.  So they're now legal keywords in a lot more places.

File: src/EditorFeatures/CSharpTest2/Recommendations/ExtensionKeywordRecommenderTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/ExtensionKeywordRecommenderTests.cs:13]: new tests that that ensure that 'extension' shows up where we expect.

File: src/EditorFeatures/CSharpTest2/Recommendations/ForKeywordRecommenderTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/ForKeywordRecommenderTests.cs:174]: new tests that validate `for` shows up in teh right place in an extension.

File: src/EditorFeatures/CSharpTest2/Recommendations/WhereKeywordRecommenderTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhereKeywordRecommenderTests.cs:636]: ensuring 'where' shows up in generic extensions where expected.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs:3389]: this is legal syntax in C#12.  so if we're taking the route of 'record' we need to ensure it parses the same prior to the version where we add extension-type support.  We can take the break once we are in C# 13 (like how record breaks if you're on the right version of C#).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1587]: i'm willing to do this @jcouv .  will do in a followup though.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1587]: Thanks much!
There's another PROTOTYPE comment for improving parsing below, if you have time: "// PROTOTYPE consider improving error recovery for `extension explicit` and `extension implicit`" (wrong order)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1585]: Braces please
```suggestion
            {
                paramList = AddError(paramList, ErrorCode.ERR_ExtensionPrimaryConstructor);
            {
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2242]: This should probably use `IsCompatBreakingFeatureEnabled` to signal that we're taking an intentional break.
In terms of documentation, I think it'll be covered when we add the logic to disallow types named "extension" (which is tracked in the test plan).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1587]: yup.  happy to take that.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1585]: :'(  added.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: done!

**Review by jcouv — APPROVED** [review]
Compiler changes LGTM Thanks (iteration 33)

**Comment by jcouv** [comment]
Tagging @AlekseyTs @jjonescz for second review of parser changes for extensions. Thanks
@CyrusNajmabadi Kindly reminder to squash since compiler-affecting change. Much appreciate

**Review by akhera99 — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1586]: >paramList = AddError(paramList, ErrorCode.ERR_ExtensionPrimaryConstructor); [](http://example.com/codeflow?start=16&length=75)

I an somewhat surprised to see that we report this error in the parser. I thought our current approach is to delay error reporting to later phases whenever possible.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1750]: Similar comment about reporting errors too early

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:3514]: >if (IsCompatBreakingFeatureEnabled(MessageID.IDS_FeatureExtensions)) [](http://example.com/codeflow?start=20&length=68)

Why not look at the next token though? I think `implicit extension.operator X(` should be parsed as an operator regardless of language version. I would prefer us to explicitly target scenarios that we explicitly decided to parse as type declarations rather than changing parsing based on language version across the board

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1581]: >ParseParenthesizedParameterList() [](http://example.com/codeflow?start=18&length=33)

Did we check binding code to make sure it doesn't try to create a symbol for the primary constructor for extension types?

File: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:2358]: >// (3,21): error CS9329: Extension cannot have a primary constructor. [](http://example.com/codeflow?start=12&length=69)

Please verify that constructor is not added as a member to the type.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1586]: It looks like for interface we check that somewhere else and we should detect extension types there as well. Also, it looks like we report `ERR_UnexpectedParameterList` for interfaces, I think that error is good enough for extension types as well. Consider removing the new error.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2244]: >IsCompatBreakingFeatureEnabled [](http://example.com/codeflow?start=27&length=30)

It doesn't look like we have this helper in main, how is it different from `IsFeatureEnabled`? If we are planning to merge this code to main, I am not certain that would be appropriate if the helper assumes we have some support for "breaking changes" feature.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2244]: >return IsCompatBreakingFeatureEnabled(MessageID.IDS_FeatureExtensions); [](http://example.com/codeflow?start=20&length=71)

As I mentioned in other similar place. I am not comfortable with language version based approach.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 58), parsing tests are not looked at

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: sounds good.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Sounds good.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Done.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: moved off of this.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: sgtm.

File: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:2358]: added check.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1581]: test added for this.

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:769]: >MessageID.IDS_FeatureExtensions.CheckFeatureAvailability(diagnostics, node, node.Keyword.GetLocation()); [](http://example.com/codeflow?start=20&length=104)

It feels like we should check for feature support regardless of the presence of `ParameterList`

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:769]: Was following the pattern above. Looks like we do the feature check only for trees that don't have specifically illegal Syntax.  

Personally, this seems fine to me.  If you have the PC, you are told it is illegal.  You remove it, and you find out you might need to upgrade your lang version.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:769]: > Was following the pattern above

What pattern? For interfaces primary constructors are not supported regardless of language version. For classes and structs the version is checked unconditionally. There is no direct similarity with extensions, for which the check for parameter list and language version are orthogonal.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:769]: For extensions, primary constructors are never allowed either. So that check happens unilaterally.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:769]: > So that check happens unilaterally. 

An I am not suggesting to change that. The checks are unrelated, and, therefore, should be performed independently. Please change the code as suggested. To avoid any confusion this is the code I expect to see:
```
            else if (node.Kind() is SyntaxKind.ExtensionDeclaration)
            {
                if (node.ParameterList != null)
                {
                    diagnostics.Add(ErrorCode.ERR_UnexpectedParameterList, node.ParameterList.GetLocation());
                }

                MessageID.IDS_FeatureExtensions.CheckFeatureAvailability(diagnostics, node, node.Keyword.GetLocation());
            }
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:3505]: >this.PeekToken(1).Kind is not SyntaxKind.ColonColonToken and not SyntaxKind.DotToken and not SyntaxKind.LessThanToken [](http://example.com/codeflow?start=24&length=117)

It looks like this condition is used in two places and we want them to stay similar as the code evolves. Consider extracting a shared helper with a meaningful name.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs:75]: >Regular11 [](http://example.com/codeflow?start=45&length=9)

@jcouv It looks like we should switch to `C# 12` here and in other similar places. I am not suggesting to do this in context of this PR.

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:769]: wfm

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs:3057]: >UsingTreeWithCSharpNext(text); [](http://example.com/codeflow?start=8&length=30)

It looks like the comment above has been addressed

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs:3172]: >UsingTreeWithCSharpNext(text); [](http://example.com/codeflow?start=8&length=30)

It looks like the comment above has been addressed

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs:3330]: >// In C# next, we always treat `explicit extension` as the start of an extension type. [](http://example.com/codeflow?start=8&length=86)

It looks like the comment is no longer relevant

**Comment by AlekseyTs** [comment]
Done with review pass (commit 65)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs]: done!

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs:3163]: done.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ExtensionParsingTests.cs:3052]: done.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler for another pass.  tnx.

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:724]: >declFlags |= SingleTypeDeclaration.TypeDeclarationFlags.IsExplicitExtension; [](http://example.com/codeflow?start=16&length=76)

I am not sure if the optimization is worth it, given that we are passing `declFlags` to `GetNonTypeMemberNames` which technically might look at the values. Therefore, the code becomes somewhat fragile long term. It looks, however, that `GetNonTypeMemberNames` doesn't actually check value of incoming `declFlags` today. Therefore, if we want to keep the optimization, we could change that parameter to an `out` parameter to enforce the fact that no one looks at the decl flags until they are fully populated.

File: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:150]: >VerifyEmitDiagnostics [](http://example.com/codeflow?start=13&length=21)

For language version testing we usually trying to cover three scenarios:
- Previous version, expected to complain about the version
- TestOptions.RegularNext - expected to not complain
- TestOptions.RegularPreview - expected to not complain

**Comment by AlekseyTs** [comment]
Done with review pass (commit 70)

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:724]: done.

File: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:150]: done.

**Review by AlekseyTs — APPROVED** [review]
Changes under Compilers LGTM (commit 72), assuming CI is passing

**Pull request merged**
